File: Problems/78/main.cc
#include <vector>
#include <iostream>
#include <string>

using namespace std;

class Solution
{
public:
    vector<vector<int>> subsets(vector<int> &nums)
    {
        vector<vector<int>> res;
        vector<int> tmp;
        generate(res, tmp, nums, 0);
        return res;
    }

    void generate(vector<vector<int>> &res, vector<int> &tmp, vector<int> &nums, int start)
    {
        res.push_back(tmp);
        for (int i = start; i < nums.size(); i++)
        {
            tmp.push_back(nums[i]);
            generate(res, tmp, nums, i + 1);
            tmp.pop_back();
        }
    }
};

File: Problems/78/main.go
package main

func main() {}

func subsets(nums []int) [][]int {
	buf := make([]int, 0, len(nums))
	return backtrack(nums, buf, 0)
}

func backtrack(nums, buf []int, start int) [][]int {
	res := make([][]int, 0, 1)
	tmp := make([]int, len(buf))
	copy(tmp, buf)
	res = append(res, tmp)
	for i := start; i < len(nums); i++ {
		buf = append(buf, nums[i])
		res = append(res, backtrack(nums, buf, i+1)...)
		buf = buf[:len(buf)-1]
	}
	return res
}
